### ISSUE TITLE
Multi-Node Multi-GPU post-training on DGXC with run:ai

### ISSUE THREAD

**Issue by otlegcah** [NVIDIA/Cosmos#100]
Hello,

I am trying to run the provided post-training demo of diffusion model on 4 nodes with 2 NVIDIA-H100-80GB GPUs each, using torchrun on DGX Cloud, and through run:ai UI, and I am having trouble with the setup.

I made the following changes to the recipe for cosmos_diffusion_7b_text2world_finetune:

Trainer setup
recipe.trainer.num_nodes = 4
recipe.trainer.devices = 2

Tensor / Sequence parallelism
recipe.trainer.strategy.tensor_model_parallel_size = 2
recipe.trainer.strategy.pipeline_model_parallel_size = 1

Data setup
recipe.data.micro_batch_size = 1
recipe.data.global_batch_size = 4

Additionally, I turned off FSDP as suggested in previous issue #97:

FSDP
recipe.trainer.strategy.ddp.with_megatron_fsdp_code_path = False
// recipe.trainer.strategy.ddp.data_parallel_sharding_strategy = "MODEL_AND_OPTIMIZER_STATES"
recipe.trainer.strategy.ddp.overlap_param_gather = False
recipe.trainer.strategy.ddp.overlap_grad_reduce = False

I launched it through run:ai UI as distributed pytorch training with the following command:

torchrun --nnodes=4 --nproc_per_node=2 cosmos1/models/diffusion/nemo/post_training/general.py --yes --factory cosmos_diffusion_7b_text2world_finetune data.path="./cached_data"

necessary environment variables set in a same way on all 3 workers and a master, each requesting 2 GPUs. Preprocessed data resides on a PVC and should be accessible from all nodes. I am using the newest image for Cosmos from NGC: nvcr.io/nvidia/nemo:cosmos.1.0.1.

I cannot make the above work, and I suppose I might have two distinct problems:
1) I am not sure about the settings in the recipe. Are the above changes to the recipe correct to enable distributed post-training demo on 4 nodes with 2 NVIDIA-H100-80GB GPUs?
2) I am not sure if my approach regarding DGXC workflow is correct. Is it possible to launch a distributed post-training demo through run:ai with NeMo-Run using the newest Cosmos image in a way described above?

From what I was able to get to know, my issues could potentially be related to:
- https://github.com/kubeflow/trainer/issues/1836
- https://github.com/NVIDIA/NeMo-Run/commit/463d6e1aae9407cbdeef1f2f035fe5cf5b2870cc
- https://github.com/NVIDIA/Cosmos/issues/53#issuecomment-2593939973

**Comment by ethanhe42** [comment]
Hi @otlegcah did you get any error? could you share the log?

**Comment by otlegcah** [comment]
Hi @ethanhe42, yes, I got an error regarding cached_data from DataLoader (see logs.txt, i.e. logs from master pod from run:ai). I could not run 4x2 now, so I attach logs from 2x2 run. The error with the new image is the same for both cases.

[Cosmos.zip](https://github.com/user-attachments/files/18759541/Cosmos.zip)

**Comment by otlegcah** [comment]
Hi @ethanhe42, yes thanks, after passing in an absolute path I was able to run the training on 2 nodes x 2 GPUs, which surprised me actually as I thought 8 GPUs were needed. Now the only thing remaining is that I am not certain, if the model improves enough, if trained this way. The learning rate seems to be quite low, and only one epoch is executed, plus it is quite slow I suppose, as it took around 5 hrs. Would you be able to give some advice on how to adjust the recipe such that the training in this configuration performs better? I attach the logs from the run for reference as well.

[Cosmos-2x2.zip](https://github.com/user-attachments/files/18774044/Cosmos-2x2.zip)

![Image](https://github.com/user-attachments/assets/79cd2b61-63ae-41c7-b624-d8f083b3b9c5)

**Comment by ethanhe42** [comment]
training hyper-parameters depend on your data. What we provided is only for the example Single Subject Post-training task.

you can expect the model adapting to the example data after post-training. 

input: https://huggingface.co/datasets/nvidia/Cosmos-NeMo-Assets?image-viewer=video-0-2B798521A2994E3FB2032A0E2CE73E00C5533667

output: https://github.com/NVIDIA/Cosmos/tree/main/cosmos1/models/diffusion/nemo/inference#example-output 

<video src="https://github.com/user-attachments/assets/a2b5bc7e-4e0a-4514-a04e-919281cee6fa">
  Your browser does not support the video tag.
</video>

**Comment by otlegcah** [comment]
The logs were based on training using this demo dataset. I now run the inference with those post-trained checkpoints and the model did adapt. Thanks for help :)

**Comment by abhigoku10** [comment]
@otlegcah  good to hear that you were able to train it successful can you please certain details on this 
1. what is the amount of dataset you used to for post training of the model 
2. how much time did it take to train the model 
thanks in advance
